toolbox/deps_update/main.go

```diff
@@ -0,0 +1,230 @@
+			if err := updateDepFile(dep); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Delete the local git repo just cloned
+func cleanUp(repo string) error {
+	if err := os.Chdir(".."); err != nil {
+		return err
+	}
+	return os.RemoveAll(repo)
+}
+
+// Update the given repository so that it uses the latest dependency references
+// push new branch to remote, create pull request on master,
+// which is auto-merged after presumbit
+func updateDependenciesOf(repo string) error {
+	if err := os.RemoveAll(repo); err != nil {
+		return err
+	}
+	if _, err := util.Shell("git clone " + githubClnt.remote(repo)); err != nil {
+		return err
+	}
+	if err := os.Chdir(repo); err != nil {
+		return err
+	}
+	defer func() {
+		if err := cleanUp(repo); err != nil {
+			log.Fatalf("Error during clean up: %v\n", err)
+		}
+	}()
+
+	if _, err := util.Shell("git checkout deps"); err != nil {
+		return err
+	}
+
+	deps, err := deserializeDeps(istioDepsFile)
+	if err != nil {
+		return err
+	}
+	depVersions, err := fingerPrintAndUpdateDepSHA(repo, &deps)
+	if err != nil {
+		return err
+	}
+	branch := "autoUpdateDeps_" + depVersions
+	exists, err := githubClnt.existBranch(repo, branch)
+	if err != nil {
+		return err
+	}
+	if exists {
+		log.Printf("Branch already exists")
+	}
+	// if branch exists, stop here and do not create another PR of identical delta
+	if err = githubClnt.closeFailedPullRequests(repo); exists || err != nil {
+		return err
+	}
+	if _, err := util.Shell("git checkout -b " + branch); err != nil {
+		return err
+	}
+	if err := updateDeps(repo, deps); err != nil {
+		return err
+	}
+	if err := serializeDeps(istioDepsFile, &deps); err != nil {
+		return err
+	}
+	if _, err := util.Shell("git add *"); err != nil {
+		return err
+	}
+	if _, err := util.Shell("git commit -m Update_Dependencies"); err != nil {
+		return err
+	}
+	if _, err := util.Shell("git push --set-upstream origin " + branch); err != nil {
+		return err
+	}
+	if err := githubClnt.createPullRequest(branch, repo); err != nil {
+		return err
+	}
+	return nil
+}
+
+func main() {
+	flag.Parse()
+	if *tokenFile == "" {
+		log.Panicf("token_file not provided\n")
+		return
+	}
+	token, err := util.GetAPITokenFromFile(*tokenFile)
+	if err != nil {
+		log.Panicf("Error accessing user supplied token_file: %v\n", err)
+	}
+	githubClnt, err = newGithubClient(*owner, token)
+	if err != nil {
+		log.Panicf("Error when initializing github client: %v\n", err)
+	}
+	if *repo != "" { // only update dependencies of this repo
+		if err := updateDependenciesOf(*repo); err != nil {
+			log.Panicf("Failed to udpate dependency: %v\n", err)
+		}
+	} else { // update dependencies of all repos in the istio project
+		repos, err := githubClnt.listRepos()
+		if err != nil {
+			log.Panicf("Error when fetching list of repos: %v\n", err)
+			return
+		}
+		for _, r := range repos {
+			if err := updateDependenciesOf(r); err != nil {
+				log.Panicf("Failed to udpate dependency: %v\n", err)
```

Comment:
Careful using panic, you may still want to update other repos even the first one hit an error